libbeat/processors/add_id/generator/es_generator.go

```diff
@@ -0,0 +1,140 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package generator
+
+import (
+	"encoding/base64"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type esTimeBasedUUIDGenerator struct{}
+
+// Singleton instance
+var _esTimeBasedUUIDGenerator IDGenerator = (*esTimeBasedUUIDGenerator)(nil)
+
+// ESTimeBasedUUIDGenerator returns the singleton instance for this generator
+func ESTimeBasedUUIDGenerator() IDGenerator {
+	return _esTimeBasedUUIDGenerator
+}
+
+var (
+	sequenceNumber uint32
+	lastTimestamp  uint64
+	delta          uint64
+	once           sync.Once
+	mac            []byte
+	mu             sync.Mutex
+)
+
+// NextID returns a base64-encoded, randomly-generated, but roughly ordered (over time), unique
+// ID. The algorithm used to generate the ID is the same as used by Elasticsearch.
+// See https://github.com/elastic/elasticsearch/blob/a666fb2266/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java
+func (*esTimeBasedUUIDGenerator) NextID() string {
+	// Initialize sequence number and mac address. We do this here instead of doing it in a package-level
+	// init function to give the runtime time to generate enough entropy for randomization.
+	initOnce()
+
+	ts, seq := nextIDData()
+	var uuidBytes [15]byte
+
+	packID(uuidBytes[:], ts, seq)
+	return base64.RawURLEncoding.EncodeToString(uuidBytes[:])
+}
+
+func initOnce() {
+	once.Do(func() {
+		sequenceNumber = rand.Uint32()
+		m, err := getSecureMungedMACAddress()
+		if err != nil {
+			panic(err)
+		}
+		mac = m
+	})
+}
+
+func nextIDData() (uint64, uint32) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	sequenceNumber++
+
+	// We only use bottom 3 bytes for the sequence number.
+	s := sequenceNumber & 0xffffff
+
+	t := timestamp()
+	return t, s
+}
+
+// timestamp returns the next timestamp to use, while accounting for system time going
+// backwards (e.g. due to a DST change).
+func timestamp() uint64 {
+	now := uint64(time.Now().UnixNano() / 1000)
+	if lastTimestamp == 0 {
+		// Last timestamp has not been previously initialized
+		lastTimestamp = now
+		return lastTimestamp
+	}
+
+	// Normally now should be later than lastTimestamp, but if the system time went backwards
+	// (e.g. due to DST change), we should compute a delta to account for this change, so we always return
+	// a value that's greater than the last call to this function.
+	if now < lastTimestamp {
+		delta = lastTimestamp - now + 1
+		lastTimestamp = now
+		return now + delta
+	}
+
+	// If the system time was reset, reset delta as well
+	if now-lastTimestamp >= delta {
+		delta = 0
+	}
+
+	lastTimestamp = now
+	return lastTimestamp
+}
```

Comment:
My worry about this function is correctness and testing. By splitting the normalization into a separate function we can add some tests to see if this really gives us monotonic IDs.

Alternatively one could try to access the monotonic clock like shown here: https://github.com/gavv/monotime/blob/master/monotime.go